sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,59 @@
 #include "sof-audio.h"
 #include "ops.h"
 
+static void sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+{
+	struct sof_ipc_free ipc_free = {
+		.hdr = {
+			.size = sizeof(ipc_free),
+			.cmd = SOF_IPC_GLB_TPLG_MSG,
+		},
+		.id = swidget->comp_id,
+	};
+	struct sof_ipc_reply reply;
+	int ret;
+
+	if (!swidget->private)
+		return;
+
+	switch (swidget->id) {
+	case snd_soc_dapm_scheduler:
+		ipc_free.hdr.cmd |= SOF_IPC_TPLG_PIPE_FREE;
+		break;
+	case snd_soc_dapm_buffer:
+		ipc_free.hdr.cmd |= SOF_IPC_TPLG_BUFFER_FREE;
+		break;
+	default:
+		ipc_free.hdr.cmd |= SOF_IPC_TPLG_COMP_FREE;
+		break;
+	}
+
+	ret = sof_ipc_tx_message(sdev->ipc, ipc_free.hdr.cmd, &ipc_free, sizeof(ipc_free),
+				 &reply, sizeof(reply));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to free widget %s\n", swidget->widget->name);
+	} else {
+		swidget->complete = 0;
+		dev_dbg(sdev->dev, "widget %s freed\n", swidget->widget->name);
+	}
+}
+
+/*
+ * Free all widgets. This function should be called after topology parsing is complete to
+ * facilitate dynamic widget loading/unloading when a PCM stream is started/stopped.
+ */
+void sof_widgets_free_all(struct snd_soc_component *scomp)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_widget *swidget;
+
+	/*
+	 * Free as many as possible even in case of errors
+	 */
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list)
+		sof_widget_free(sdev, swidget);
+}
+
```

Comment:
Humm, what happens if some of the widgets are part of a graph that doesn't include any PCM devices.

Think for example of a firmware that would take audio input, filter it and play it back (typical flow for guitar effects).

Or a codec-to-codec processing. 

Or a voice call without any PCM active (this is very real, this was enabled on phones).

I think you can really setup pipelines on demand if they are connected to a PCM device. That's not necessarily the general case. Think e.g. of the tone control. What would happen there?

Reply:
@plbossart for usecases that dont have a PCM, we'd still do a hw_params/hw_free isnt it?

Reply:
I don't think so. If there is no PCM, there is no hw_params to setup...

Reply:
What I meant was even without hw_params, in order to trigger the tone pipeline for example, we'd need to pass the pipeline params. This would be a speacial case though and without a path from host->dai, the DAPM status of the widgets wont actually be useful in setting up the pipelines. I am not sure how to handle this case. Do you have a suggestion?

Reply:
when I played with the tone pipeline 3 years ago (seems like ages ago), I had to mess with a virtual PCM to update the codec DAPM power status. I might have to revisit that and see how it actually worked